flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -0,0 +1,298 @@
+
+	/** Logger for this class. */
+	private static final Logger LOG = LoggerFactory.getLogger(TaskLocalStateStoreImpl.class);
+
+	/** Maximum number of retained snapshots. */
+	@VisibleForTesting
+	static final int MAX_RETAINED_SNAPSHOTS = 5;
+
+	/** Dummy value to use instead of null to satisfy {@link ConcurrentHashMap}. */
+	private final TaskStateSnapshot NULL_DUMMY = new TaskStateSnapshot();
+
+	/** JobID from the owning subtask. */
+	@Nonnull
+	private final JobID jobID;
+
+	/** AllocationID of the owning slot. */
+	@Nonnull
+	private final AllocationID allocationID;
+
+	/** JobVertexID of the owning subtask. */
+	@Nonnull
+	private final JobVertexID jobVertexID;
+
+	/** Subtask index of the owning subtask. */
+	@Nonnegative
+	private final int subtaskIndex;
+
+	/** The configured mode for local recovery. */
+	@Nonnull
+	private final LocalRecoveryConfig localRecoveryConfig;
+
+	/** Executor that runs the discarding of released state objects. */
+	@Nonnull
+	private final Executor discardExecutor;
+
+	/** Lock for synchronisation on the storage map and the discarded status. */
+	@Nonnull
+	private final Object lock;
+
+	/** Status flag if this store was already discarded. */
+	@GuardedBy("lock")
+	private boolean discarded;
+
+	/** Maps checkpoint ids to local TaskStateSnapshots. */
+	@Nonnull
+	@GuardedBy("lock")
+	private final SortedMap<Long, TaskStateSnapshot> storedTaskStateByCheckpointID;
+
+	public TaskLocalStateStoreImpl(
+		@Nonnull JobID jobID,
+		@Nonnull AllocationID allocationID,
+		@Nonnull JobVertexID jobVertexID,
+		@Nonnegative int subtaskIndex,
+		@Nonnull LocalRecoveryConfig localRecoveryConfig,
+		@Nonnull Executor discardExecutor) {
+
+		this.jobID = jobID;
+		this.allocationID = allocationID;
+		this.jobVertexID = jobVertexID;
+		this.subtaskIndex = subtaskIndex;
+		this.discardExecutor = discardExecutor;
+		this.lock = new Object();
+		this.storedTaskStateByCheckpointID = new TreeMap<>();
+		this.discarded = false;
+		this.localRecoveryConfig = localRecoveryConfig;
+	}
+
+	@Override
+	public void storeLocalState(
+		@Nonnegative long checkpointId,
+		@Nullable TaskStateSnapshot localState) {
+
+		if (localState == null) {
+			localState = NULL_DUMMY;
+		}
+
+		LOG.info("Storing local state for checkpoint {}.", checkpointId);
+		LOG.debug("Local state for checkpoint {} is {}.", checkpointId, localState);
+
+		Map<Long, TaskStateSnapshot> toDiscard = new HashMap<>(MAX_RETAINED_SNAPSHOTS);
+
+		synchronized (lock) {
+			if (discarded) {
+				// we ignore late stores and simply discard the state.
+				toDiscard.put(checkpointId, localState);
+			} else {
+				TaskStateSnapshot previous =
+					storedTaskStateByCheckpointID.put(checkpointId, localState);
+
+				if (previous != null) {
+					toDiscard.put(checkpointId, previous);
+				}
+
+				// remove from history.
+				while (storedTaskStateByCheckpointID.size() > MAX_RETAINED_SNAPSHOTS) {
+					Long removeCheckpointID = storedTaskStateByCheckpointID.firstKey();
```

Comment:
I found this can be problematic, even this method has been invoked but it doesn't mean the checkpoint is truly completed. If we perform remove local task state here we may removed the `really completed local checkpoint data`. Image the situation when `max concurrent checkpoint number = MAX_RETAINED_SNAPSHOTS + 1`...

Reply:
If we have to remove local completed checkpoint here, I think `MAX_RETAINED_SNAPSHOTS` should bigger than `max concurrent checkpoint number.`

Reply:
Yes, this pruning can be removed without replacement.

Reply:
Aha, I think simply removed this pruning will also lead to problematic.... Imagine that, if checkpoint always completed in this subtask but failed on global, this will lead to resources leak. I think this can be fixed by introduce a scheduled `canceller` for clear these `pending completed checkpoint` just like what we have done in `CheckpointCoordinator`.

Reply:
I don't think that should matter too much, the data will already be deleted once a checkpoint completes.

Reply:
Yes, once a checkpoint completed, all `pending completed checkpoint` will be deleted, but if the job meet a problem and can never get a `global completed checkpoint` these will lead to the problems that I described. But I agree with you this doesn't matter too much, this is just a extreme case.

Reply:
Yes, and that was the original intention of having a limit. I still think that the limit might be the most proper way of doing it, but it has to be aligned with the maximum number of concurrent checkpoints, as you pointed out. I could reintroduce this optimization in a followup.